### PR TITLE
Implement Memory64 support for AOT

### PR DESCRIPTION
**Pull request by loganek**
All spec tests are passing now; I also ran the memory64 tests from @TianlongLiang 's  wamr-test repository.

### CODE REVIEW

**Comment by wenyongh** [comment]
@loganek the macos CI failure should have been fixed with PR https://github.com/bytecodealliance/wasm-micro-runtime/pull/3366, please rebase your code to run again. Thanks.

**Comment by loganek** [comment]
@wenyongh thanks for letting me know, I'll do that. The pr is in the draft state yet so I fix the ci failures and do a self review; I'll switch it to non-draft once it's ready for the review

**Comment by loganek** [comment]
@wenyongh / @TianlongLiang I think the code is ready for review, I'd appreciate your feedback.

File: core/iwasm/aot/aot_loader.c [hunk 1]
```diff
@@ -3632,6 +3632,21 @@ load_relocation_section(const uint8 *buf, const uint8 *buf_end,
...
+    /* TODO: multi-memories for now assuming the memory idx type is consistent
+     * across multi-memories */
+    if (module->import_memory_count > 0)
+        return !!(module->import_memories[0].memory_flags & MEMORY64_FLAG);
+    else if (module->memory_count > 0)
+        return !!(module->memories[0].memory_flags & MEMORY64_FLAG);
+
+    return false;
+}
+#endif
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:3648]: In check_feature_flags function of this file, we check whether the related feature is enabled when the feature flag is set in the AOT file. Had better also emit the memory64 feature flag in the AOT file when there is a 64-bit memory in the wasm file, and add the check in check_feature_flags.

File: tests/wamr-test-suites/spec-test-script/runtest.py [hunk 2]
```diff
@@ -1119,9 +1119,6 @@ def compile_wasm_to_aot(wasm_tempfile, aot_tempfile, runner, opts, r, output = '
         cmd.append("--enable-gc")
         cmd.append("--enable-tail-call")
 
-    if opts.memory64:
-        cmd.append("--enable-memory64")
```
Comment (wenyongh) [inline tests/wamr-test-suites/spec-test-script/runtest.py:1123]: Seems we don't need to add `--enable-memory64` flag for wamrc, it can auto generate the correct aot file according to the memory definition in the wasm file?

**Comment by TianlongLiang** [comment]
Hi, @loganek, I think maybe it's worth considering 
1. adding a command line option like `--enable-memory64` in wamrc so that user can opt it out like user can choose to `--enable-multi-thread` and `--enable-gc`, in this way, I think `tests/wamr-test-suites/spec-test-script/thread_proposal_remove_memory64_flag_case.patch` probably won't be needed 
2. emitting the corresponding feature flags(`WASM_FEATURE_MEMORY64`) in `aot_emit_aot_file.c` when there is actually i64 type memory section and correspondingly modify `check_feature_flags` in `aot_loader.c`, this way I think it's also can get rid of the patch

**Comment by wenyongh** [comment]
> @wenyongh / @TianlongLiang I think the code is ready for review, I'd appreciate your feedback.

@loganek thanks for the quick update. I just read the code quickly and added two comments, and hope to read it carefully to understand more. But since Tianlong and I are in holiday in the left of this week, we might response slowly then.

**Comment by loganek** [comment]
> But since Tianlong and I are in holiday in the left of this week, we might response slowly then.
no worries, enjoy your holiday :)

File: tests/wamr-test-suites/spec-test-script/runtest.py (hunk 2 above)
Comment (loganek) [reply tests/wamr-test-suites/spec-test-script/runtest.py:1123]: yes, that's why I removed this option.

**Comment by loganek** [comment]
> adding a command line option like --enable-memory64 in wamrc so that user can opt it out like user can choose to --enable-multi-thread and --enable-gc, in this way, I think tests/wamr-test-suites/spec-test-script/thread_proposal_remove_memory64_flag_case.patch probably won't be needed

I personally don't think this flag is needed, I'm not sure if there's usecase where the user has a memory64 module and does NOT want to build with memory64 - in all cases that should fail, so I'd rather delete the cli option and let compiler automatically detect the type of module.

The patch is to handle edge case where the module is invalid, and I don't think we actually ever want to support such case. In fact, that test was removed in the most recent version of the testsuite (I just didn't want to upgrade the whole test suite in this PR to keep changes focused to memory64 only).

File: core/iwasm/compilation/aot_emit_memory.c [hunk 3]
```diff
@@ -987,13 +1003,13 @@ aot_compile_op_memory_init(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
...
     POP_I32(offset);
-    POP_I32(dst);
+    POP_MEM_OFFSET(dst);
 
     param_types[0] = INT8_PTR_TYPE;
     param_types[1] = I32_TYPE;
     param_types[2] = I32_TYPE;
     param_types[3] = I32_TYPE;
-    param_types[4] = I32_TYPE;
+    param_types[4] = MEMORY64_COND_VALUE(I64_TYPE, I32_TYPE);
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_memory.c:1012]: The `dst` argument of `aot_memory_init` is `mem_offset_t`, which is defined as:
```C
#if WASM_ENABLE_MEMORY64 != 0
typedef uint64 mem_offset_t;
#else
typedef uint32 mem_offset_t;
#endif
```
So it is uint64 when WASM_ENABLE_MEMORY64 is set as 1, no matter the actual memory is 64 bit or 32 bit.
And here the param type of dst is decided according to the memory type, it is I32_TYPE when the memory is 32 bit. The issue reported in #3375 may occur again: aot runtime enables memory64 feature and the argument type is uint64, while memory in aot file is 32bit and aot code passes uint32 to the argument.

How about changing the param type in aot_memory_init to size_t, and here set the param type according to `comp_ctx->pointer_size`, and extend the param to uint64 if needed before passing to aot_memory_init? Similar to #3378.

File: core/iwasm/compilation/aot_emit_memory.c [hunk 4]
```diff
@@ -1165,16 +1181,16 @@ aot_compile_op_memory_fill(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
...
-    POP_I32(dst);
+    POP_MEM_OFFSET(dst);
 
     if (!(dst_addr = check_bulk_memory_overflow(comp_ctx, func_ctx, dst, len)))
         return false;
 
     param_types[0] = INT8_PTR_TYPE;
     param_types[1] = I32_TYPE;
-    param_types[2] = I32_TYPE;
+    param_types[2] = MEMORY64_COND_VALUE(I64_TYPE, I32_TYPE);
```
[the remark this answers is not in the document]
Comment (loganek) [reply core/iwasm/compilation/aot_emit_memory.c:1193]: thanks, updated

File: core/iwasm/compilation/aot_emit_memory.c [hunk 5]
```diff
@@ -1097,7 +1113,7 @@ aot_compile_op_memory_copy(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
 
         param_types[0] = INT8_PTR_TYPE;
         param_types[1] = INT8_PTR_TYPE;
-        param_types[2] = I32_TYPE;
+        param_types[2] = MEMORY64_COND_VALUE(I64_TYPE, I32_TYPE);
```
[the remark this answers is not in the document]
Comment (loganek) [reply core/iwasm/compilation/aot_emit_memory.c:1116]: right, and in fact this should be based on the target architecture, not based on the memory type.

File: core/iwasm/aot/aot_loader.c (hunk 1 above)
Comment (loganek) [reply core/iwasm/aot/aot_loader.c:3648]: hmm, I'm quite sure I've got `wasm_memory_check_flags` call in AOT (that's why I refactored it with the `is_aot` flag parameter) but looks like I must have removed it when doing some refactoring. I'll add it back in the `load_memory_info`, I don't think we need to add feature flags if the information about the memory type is already encoded in the aot bundle anyway.

File: core/iwasm/compilation/aot_emit_memory.c [hunk 6]
```diff
@@ -854,7 +869,7 @@ aot_compile_op_memory_grow(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
 
     /* Call function aot_enlarge_memory() */
     param_values[0] = func_ctx->aot_inst;
-    param_values[1] = delta;
+    param_values[1] = LLVMBuildTrunc(comp_ctx->builder, delta, I32_TYPE, "");
```
[the remark this answers is not in the document]
Comment (loganek) [reply core/iwasm/compilation/aot_emit_memory.c:872]: Good idea, updated.

File: core/iwasm/compilation/aot_emit_memory.c [hunk 7]
```diff
@@ -852,9 +885,27 @@ aot_compile_op_memory_grow(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
...
+            return false;
+        }
+        BUILD_ICMP(LLVMIntUGT, delta, u32_max, ret_value, "page_size_cmp");
+        if (!(check_page_size = LLVMAppendBasicBlockInContext(
+                  comp_ctx->context, func_ctx->func, "check_page_size"))) {
+            aot_set_last_error("llvm add basic block failed.");
+            goto fail;
+        }
+        if (!(aot_emit_exception(comp_ctx, func_ctx, EXCE_INTEGER_OVERFLOW,
+                                 true, ret_value, check_page_size)))
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_memory.c:901]: When `memory.grow` fails, it should return -1 but not throw exception, see wasm spec and the implementation in interpreter:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_interp_classic.c#L4557

Suggest here:
```C
    BUILD_ICMP(LLVMIntUGT, delta, u32_max, ret_value1, "page_size_cmp");
```

And change below to:
```C
    BUILD_ICMP(LLVMIntUGT, ret_value, I8_ZERO, ret_value, "mem_grow_ret");

    BUILD_OP(Or, ret_value, ret_value1, ret_value, "or");  => ret_value = ret_value | ret_value1;

    /* ret_value = ret_value == true ? delta : pre_page_count */
    if (!(ret_value = LLVMBuildSelect(
              comp_ctx->builder, ret_value, mem_size,
              MEMORY64_COND_VALUE(I64_NEG_ONE, I32_NEG_ONE), "mem_grow_ret"))) {
        aot_set_last_error("llvm build select failed.");
        return false;
    }

    PUSH_PAGE_COUNT(ret_value);
```

File: core/iwasm/interpreter/wasm_runtime.c [hunk 8]
```diff
@@ -4176,7 +4176,7 @@ llvm_jit_invoke_native(WASMExecEnv *exec_env, uint32 func_idx, uint32 argc,
 #if WASM_ENABLE_BULK_MEMORY != 0
 bool
 llvm_jit_memory_init(WASMModuleInstance *module_inst, uint32 seg_index,
-                     uint32 offset, uint32 len, uint32 dst)
+                     uint32 offset, uint32 len, size_t dst)
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:4179]: Should also call `clamp_u64_to_u32(memory_inst->memory_data_size - dst)` in below `bh_memcpy_s`

File: core/iwasm/compilation/aot_emit_memory.c [hunk 9]
```diff
@@ -864,13 +915,14 @@ aot_compile_op_memory_grow(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx)
     BUILD_ICMP(LLVMIntUGT, ret_value, I8_ZERO, ret_value, "mem_grow_ret");
 
     /* ret_value = ret_value == true ? delta : pre_page_count */
```
Comment (wenyongh) [inline core/iwasm/compilation/aot_emit_memory.c:917]: ret_value = ret_value == true ? pre_page_count : -1

File: core/iwasm/compilation/aot_emit_memory.c (hunk 7 above)
Comment (wenyongh) [reply core/iwasm/compilation/aot_emit_memory.c:901]: Sorry that the logic should be `if (delta <= u32_max && aot_enlarge_memory()) then return pre_page_count, else return -1`, so seems had better:
```C
BUILD_ICMP(LLVMIntULE, delta, u32_max, ret_value1, "page_size_cmp");
...

BUILD_ICMP(LLVMIntUGT, ret_value, I8_ZERO, ret_value, "mem_grow_ret");

#if WASM_ENABLE_MEMORY64 != 0
    BUILD_OP(And, ret_value1, ret_value, ret_value, "and");  => ret_value = ret_value1 & ret_value;
#endif
```

File: core/iwasm/aot/aot_runtime.c [hunk 10]
```diff
@@ -2401,8 +2445,16 @@ execute_malloc_function(AOTModuleInstance *module_inst, WASMExecEnv *exec_env,
     if (exec_env_created)
         wasm_exec_env_destroy(exec_env_created);
 
-    if (ret)
-        *p_result = argv[0];
+    if (ret) {
+#if WASM_ENABLE_MEMORY64 != 0
+        if (is_memory64)
+            *p_result = GET_I64_FROM_ADDR(&argv[0]);
```
[the remark this answers is not in the document]
Comment (loganek) [reply core/iwasm/aot/aot_runtime.c:2451]: true, also the `size` parameter should be `uint64`, updated that too.

File: core/iwasm/interpreter/wasm_runtime.c (hunk 8 above)
Comment (loganek) [reply core/iwasm/interpreter/wasm_runtime.c:4179]: I haven't looked into JIT and there's no way to test memory64 in JIT right now as memory64 is not supported in this mode, but that's a reasonable change, updated.

**Comment by loganek** [comment]
> it's rather important to fail when it's expected to fail because, in case of aot, it's wamrc's resposibility to validate the module.

I agree it's important, I don't think though that the explicit flag is needed; the module is self-explanatory in terms of the memory type. If we really want to do that explicit validation for some reason, I think we should be consistent across the whole wamr, and not only have the explicit flag in the compiler, but also for runtime (and not only for memory64, but also other features like threads). As I said, I don't see a need for that, and that'd slightly complicate user's life as they'll need to know in advance what type of module they load. As of today they can do the validation beforehand if they need it, but it's optional.

File: core/iwasm/aot/aot_runtime.c [hunk 11]
```diff
@@ -2401,24 +2441,46 @@ execute_malloc_function(AOTModuleInstance *module_inst, WASMExecEnv *exec_env,
...
+            *p_result = argv.u32[0];
+        }
+    }
     return ret;
 }
 
 static bool
 execute_free_function(AOTModuleInstance *module_inst, WASMExecEnv *exec_env,
-                      AOTFunctionInstance *free_func, uint32 offset)
+                      AOTFunctionInstance *free_func, uint64 offset)
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:2459]: Should also modify:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/ca61184ced7d885395bc2425c87f0d063935dd1e/core/iwasm/aot/aot_runtime.c#L2623-L2624
change `(uint32)ptr` to `ptr`

**Comment by wenyongh** [comment]
> > it's rather important to fail when it's expected to fail because, in case of aot, it's wamrc's resposibility to validate the module.
> 
> I agree it's important, I don't think though that the explicit flag is needed; the module is self-explanatory in terms of the memory type. If we really want to do that explicit validation for some reason, I think we should be consistent across the whole wamr, and not only have the explicit flag in the compiler, but also for interpreter (and not only for memory64, but also other features like threads). As I said, I don't see a need for that, and that'd slightly complicate user's life as they'll need to know in advance what type of module they load. As of today they can do the validation beforehand if they need it, but it's optional.

Per my understanding, wamrc supports both memory32 and memory64 and can emit correct AOT files accordingly. And for memory64, the memory info in the AOT file also has the memory64 flag (bit 2 is set), the aot runtime can report error if memory64 feature isn't enabled, which is consistent with interpreter. So no need to explicitly add `--enable-memory64` flag for wamrc, right? I think it is acceptable since it seems more convenient to developers.

**Comment by wenyongh** [comment]
@loganek the PR LGTM except one issue left (change `(uint32)ptr` to `ptr` when calling execute_free_function), could you help address it and resolve the conflicting file? I think we can merge this PR then if @yamt and @TianlongLiang are also good.

File: core/iwasm/compilation/aot_compiler.c [hunk 12]
```diff
@@ -2743,7 +2760,7 @@ aot_compile_func(AOTCompContext *comp_ctx, uint32 func_index)
                     case SIMD_v128_store:
                     {
                         read_leb_uint32(frame_ip, frame_ip_end, align);
-                        read_leb_uint32(frame_ip, frame_ip_end, offset);
+                        read_leb_mem_offset(frame_ip, frame_ip_end, offset);
```
Comment (TianlongLiang) [inline core/iwasm/compilation/aot_compiler.c:2763]: Should also replace the `read_leb_uint32(frame_ip, frame_ip_end, offset);` for L3009, L3023, L3035. 

And I think it's best to leave a comment before `case WASM_OP_SIMD_PREFIX` and `case WASM_OP_ATOMIC_PREFIX` with: `TODO: in the future, replace the offset type in aot_compile_op_xxx for threads and SIMD opcodes compilation`

Comment (loganek) [reply core/iwasm/compilation/aot_compiler.c:2763]: hmm, I think I'd rather just change this now instead of leaving todos

**Comment by loganek** [comment]
@wenyongh I've addressed all the comments, thanks a lot for the review.

Unless there are any other comments, the PR is ready for merge.

**Review by wenyongh — APPROVED** [review]
LGTM

**Comment by wenyongh** [comment]
> @wenyongh I've addressed all the comments, thanks a lot for the review.
> 
> Unless there are any other comments, the PR is ready for merge.

Great work! I think we can merge the PR if it is also good from @TianlongLiang.

**Comment by TianlongLiang** [comment]
LGTM
